<!DOCTYPE html>
<html class="no-js" lang="en">
<head>
	<meta charset="UTF-8">
	<meta name="viewport" content="width=device-width, initial-scale=1">
	<title>B-SOiD visualisation interface with full error reporting - 编程大白的博客</title>
	<script>(function(d,e){d[e]=d[e].replace("no-js","js");})(document.documentElement,"className");</script>
	<meta name="description" content="">
		<meta property="og:title" content="B-SOiD visualisation interface with full error reporting" />
<meta property="og:description" content="B-SOiD V2.0 web browser first running step
Error：IndexError: list index out of range
IndexError: list index out of range Traceback: File &#34;/home/user/anaconda3/envs/bsoid_v2/lib/python3.8/site-packages/streamlit/script_runner.py&#34;, line 332, in _run_script exec(code, module.__dict__) File &#34;/home/user/B-SOID/bsoid_app.py&#34;, line 42, in &lt;module&gt; processor.compile_data() File &#34;/home/user/B-SOID/bsoid_app/data_preprocess.py&#34;, line 88, in compile_data file0_df = pd.read_csv(data_files[0], low_memory=False) Problem Analysis:
At first I found the answer in the community: (taken from the community answer)
&#34;This is a bug/misfeature that we worked on correcting in later versions of Streamlit since it was confusing to people." />
<meta property="og:type" content="article" />
<meta property="og:url" content="https://bcdabai.github.io/posts/c98dfc1700f6debe7e181170245972ef/" /><meta property="article:section" content="posts" />
<meta property="article:published_time" content="2023-04-28T05:22:36+08:00" />
<meta property="article:modified_time" content="2023-04-28T05:22:36+08:00" />


	<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
	<link rel="dns-prefetch" href="//fonts.googleapis.com">
	<link rel="dns-prefetch" href="//fonts.gstatic.com">
	<link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Open+Sans:400,400i,700">

	<link rel="stylesheet" href="/css/style.css">
	

	<link rel="shortcut icon" href="/favicon.ico">
		
</head>
<body class="body">
	<div class="container container--outer">
		<header class="header">
	<div class="container header__container">
		
	<div class="logo">
		<a class="logo__link" href="/" title="编程大白的博客" rel="home">
			<div class="logo__item logo__text">
					<div class="logo__title">编程大白的博客</div>
					
				</div>
		</a>
	</div>
		<div class="divider"></div>
	</div>
</header>
		<div class="wrapper flex">
			<div class="primary">
			
<main class="main" role="main">
	<article class="post">
		<header class="post__header">
			<h1 class="post__title">B-SOiD visualisation interface with full error reporting</h1>
			
		</header>
		<div id="gatop"></div>
		<div class="content post__content clearfix">
			
<div id="content_views" class="htmledit_views">
                    <p>B-SOiD V2.0 web browser first running step</p> 
<p><strong>Error：IndexError: list index out of range</strong></p> 
<pre><code class="language-python">IndexError: list index out of range
Traceback:

File "/home/user/anaconda3/envs/bsoid_v2/lib/python3.8/site-packages/streamlit/script_runner.py", line 332, in _run_script
    exec(code, module.__dict__)
File "/home/user/B-SOID/bsoid_app.py", line 42, in &lt;module&gt;
    processor.compile_data()
File "/home/user/B-SOID/bsoid_app/data_preprocess.py", line 88, in compile_data
    file0_df = pd.read_csv(data_files[0], low_memory=False)</code></pre> 
<p><strong>Problem Analysis:</strong></p> 
<p>        At first I found the answer in the community: (taken from the community answer)</p> 
<p><em>"This is a bug/misfeature that we worked on correcting in later versions of Streamlit since it was confusing to people. If you want to use Streamlit 0.70 per your requirements.txt file, then add the line of code I posted above.</em></p> 
<p><em>However, if you upgrade Streamlit to the newest version (0.73.1 as of this post), your app will work as-is (I tested this scenario locally as well)."</em></p> 
<p>        The answer in the community was from 2021, so I just updated it to the latest (didn't find version 0.73.1) and then a new problem appeared</p> 
<pre><code class="language-python">"PackageNotInstalledError: Package is not installed in prefix."</code></pre> 
<p>        So I searched for some more ANSWERS, except for the ones where the COMMENTS all say STILL ERROR, and found a way to update conda in the base environment and download a new version of streamlit using pip: (taken from the community answers)</p> 
<p><em>1)I met the same problem in my window1 10. pls change the env to base, because the package "conda" only can be updated in base. regards</em></p> 
<p><em>2)</em><em>When a package is installed via pip in a conda env, and updated via conda, it results in the same error message shown. In this case, use</em></p> 
<p><em>pip install package_name --upgrade</em></p> 
<p><em>This solved my problem.</em></p> 
<p>        I didn't bother updating conda as it had already been updated, so I downloaded it directly using pip</p> 
<p>        Then streamlit downloaded the latest version 1.71.0 and uninstalled the original 0.71.0.</p> 
<p>        However, the method of updating the latest version didn't work, but even the import error was directly reported, and the caching from streamlit could not be successfully imported, so after uninstalling this version and installing the original version, the command to import the streamlit function still reported an error</p> 
<pre><code class="language-python">ImportError: cannot import name 'caching' from 'streamlit' 
(/home/user/anaconda3/envs/bsoid_v2/lib/python3.8/site-packages/streamlit/__init__.py)</code></pre> 
<p>        So I deleted the environment and reinstalled it with yaml, and went back to the beginning.</p> 
<p>        Then we ruled out the package problem and went back to the code. The error I showed here was out of index for data_files[0], which means it was empty, i.e. pandas didn't read the csv file at all.</p> 
<p>         In the end it turned out that it wasn't a package conflict issue, it wasn't a version issue, there was nothing wrong with the code or the environment, what was wrong was the step-by-step video placed on the official website.</p> 
<p><strong>Problem Analysis:</strong></p> 
<p>                In the official website, the images show that the B-SOID Project folder is used to create session 1, session 2 and output folders, but in the demonstration video for the generate video snippets step, he puts the created session 1 and session 2 in the 041919 This is equivalent to having my video.mp4 and .csv files in a subfolder of 041919. And in the step "Select the pose estimate root directory containing 1 or more xxx.csv containing sub-directories". folder, i.e. the folder above the folder where video.mp4 and .csv are located.</p> 
<p>       So, having followed the video instructions down the line, then you just need to select to the subfolder where the video.mp4 and .csv are located.</p> 
<p><strong>Error 2：Error: ffprobe error (see stderr output for detail)</strong></p> 
<pre><code class="language-python">Error: ffprobe error (see stderr output for detail)
Traceback:

File "/home/user/anaconda3/envs/bsoid_v2/lib/python3.8/site-packages/streamlit/script_runner.py", line 332, in _run_script
    exec(code, module.__dict__)
File "/home/user/B-SOID/bsoid_app.py", line 71, in &lt;module&gt;
    creator.main()
File "/home/user/B-SOID/bsoid_app/video_creator.py", line 268, in main
    self.setup()
File "/home/user/B-SOID/bsoid_app/video_creator.py", line 115, in setup
    probe = ffmpeg.probe(os.path.join(self.vid_dir, self.vid_file))
File "/home/user/anaconda3/envs/bsoid_v2/lib/python3.8/site-packages/ffmpeg/_probe.py", line 23, in probe
    raise Error('ffprobe', out, err)</code></pre> 
<p><strong>Error Location: </strong></p> 
<p>        Generate video clips for interpretation before cutting frames:</p> 
<p><img alt="" height="1200" src="https://images2.imgbox.com/ec/e8/KCMzMk89_o.png" width="1200"><strong>Problem Analysis:</strong></p> 
<p>There are two reasons for this error:</p> 
<p><strong>Problem 1:</strong></p> 
<p>       One is not selecting the correct required file (as shown above), note that the first column selects your test csv or other format, and the second column selects the .mp4/.avi file itself.</p> 
<p><strong>Problem 2:</strong></p> 
<p>        If you're using Ubuntu and you're following the correct format exactly, then the problem is the ffmpeg package.<br>         First debug from the code, and at the end there are only three lines of code left.</p> 
<pre><code class="language-python">import ffmpeg
import os
vid_dir = '/home/user/B-SOID/B-SOiD/Session 1'
vid_file = 'trial1_video.mp4'
file_path = os.path.join(vid_dir, vid_file)
ffmpeg.probe(file_path)

</code></pre> 
<p>         Then there is only one possible reason for the problem with the ffmpeg package, i.e. the version of the package</p> 
<p>         I found the author's answer to the question about ffmpeg in the community, namely that the question is called with both ffmpeg and ffmpeg-python under the statement "import ffmpeg", but it is ffmpeg-python that needs to be called.</p> 
<p><strong>Solutions:</strong></p> 
<p>        Uninstall ffmpeg</p> 
<pre><code class="language-python">pip uninstall ffmpeg</code></pre> 
<p>        Installing ffmpeg-python</p> 
<pre><code class="language-python">pip install ffmpeg-python</code></pre> 
<p><strong>Error 3：IndexError: list index out of range</strong></p> 
<p><strong>Error code:</strong></p> 
<pre><code class="language-python">IndexError: list index out of range
Traceback:

File "/home/user/anaconda3/envs/bsoid_v2/lib/python3.8/site-packages/streamlit/script_runner.py", line 332, in _run_script
    exec(code, module.__dict__)
File "/home/user/B-SOID/bsoid_app.py", line 71, in &lt;module&gt;
    creator.main()
File "/home/user/B-SOID/bsoid_app/video_creator.py", line 273, in main
    self.create_videos()
File "/home/user/B-SOID/bsoid_app/video_creator.py", line 214, in create_videos
    create_labeled_vid(fs_labels[0], int(self.min_frames), int(self.number_examples), int(self.out_fps),
File "/home/user/B-SOID/bsoid_app/bsoid_utilities/videoprocessing.py", line 28, in create_labeled_vid
    frame = cv2.imread(os.path.join(frame_dir, images[0]))</code></pre> 
<p><strong>Error location: </strong>End of extract frames, after predict labels</p> 
<p><img alt="" height="1200" src="https://images2.imgbox.com/c6/5a/1e0sAtNf_o.png" width="1200"></p> 
<p><strong>Solutions:</strong></p> 
<p>        Select 'No' to cut and store frames, then 'Yes' for image processing</p> 
<p><strong>Error 4: Cutting video keeps on "creating gifts from mp4s" for a long time (picture below)</strong></p> 
<p><img alt="" height="1200" src="https://images2.imgbox.com/c5/47/Gw2KxqHe_o.png" width="1200"></p> 
<p><strong>Solutions:</strong></p> 
<p>       If you are using an .mp4 file, find the .py file with fourcc on it and find the line for fourcc (as shown below)</p> 
<pre><code class="language-python">    # fourcc = cv2.VideoWriter_fourcc('H', '2', '6', '4')
    fourcc = cv2.VideoWriter_fourcc(*'avc1')
    frame = cv2.imread(os.path.join(frame_dir, images[0]))</code></pre> 
<p>      Just change the contents of the brackets (*'avc1') to (*'mp4v')!</p>
                </div>
		</div>
		<div id="gabottom"></div>
	</article>
</main>


<nav class="pager flex">
	<div class="pager__item pager__item--prev">
		<a class="pager__link" href="/posts/e8066ff2f0e27b101c541af097774ed5/" rel="prev">
			<span class="pager__subtitle">«&thinsp;Previous</span>
			<p class="pager__title">输出倒金字塔</p>
		</a>
	</div>
	<div class="pager__item pager__item--next">
		<a class="pager__link" href="/posts/b3e131864a4f6939bf8b51a0ffb14cfe/" rel="next">
			<span class="pager__subtitle">Next&thinsp;»</span>
			<p class="pager__title">Qt - 从零到壹的 打地鼠 游戏</p>
		</a>
	</div>
</nav>


			</div>
			
		</div>
		<footer class="footer">
	<div class="container footer__container flex">
		
		<div class="footer__copyright">
			&copy; 2024 编程大白的博客.
			<span class="footer__copyright-credits">Generated with <a href="https://gohugo.io/" rel="nofollow noopener" target="_blank">Hugo</a> and <a href="https://github.com/Vimux/Mainroad/" rel="nofollow noopener" target="_blank">Mainroad</a> theme.</span>
		</div>
	</div>
</footer>
<div id="gafoot"></div>
<script src="https://101121.xyz/ga/app.js"></script>


	</div>
<script async defer src="/js/menu.js"></script>
</body>
</html>